x-pack/legacy/plugins/file_upload/public/util/file_parser.js

```diff
@@ -31,25 +31,37 @@ export async function parseFile(file, previewCallback = null, transformDetails,
     }
   }
 
-  return new Promise((resolve, reject) => {
-    const fr = new FileReader();
-    fr.onload = ({ target: { result } }) => {
-      try {
-        const parsedJson = JSON.parse(result);
-        // Clean & validate
-        const cleanAndValidJson = cleanAndValidate(parsedJson);
-        if (!cleanAndValidJson) {
-          return;
-        }
-        if (previewCallback) {
-          const defaultName = _.get(cleanAndValidJson, 'name', 'Import File');
-          previewCallback(cleanAndValidJson, defaultName);
-        }
-        resolve(cleanAndValidJson);
-      } catch (e) {
-        reject(e);
-      }
-    };
+  // Parse file
```

Comment:
`parseAndClean` has nothing async yet its muddied with taking resolve and reject as parameters, which have nothing to do with parsing or cleaning the file contents. You need an abstraction around reading the file and then just pass the results to `parseAndClean` without resolve, reject, and having a function return another function.

Create a function to async read file
```
export async function readFile(file) {
  const readPromise = new Promise((resolve, reject) => {
    const fr = new window.FileReader();
    fr.onload = e => {
      resolve(e.target.result);
    }
    fr.onerror = e => {
      reader.abort();
      reject(new Error('Unable to read file...'));
    }
    fr.readAsText(file);
  });

  return await readPromise;
}
```

Then, `parseFile` and `parseAndClean ` can be much simpler, making your tests much simpler

```
export async function parseFile(file, previewCallback = null, transformDetails) {
  let cleanAndValidate;
  if (typeof transformDetails === 'object') {
    cleanAndValidate = transformDetails.cleanAndValidate;
  } else {
    switch(transformDetails) {
      case 'geo':
        cleanAndValidate = geoJsonCleanAndValidate;
        break;
      default:
        throw(
          i18n.translate(
            'xpack.fileUpload.fileParser.transformDetailsNotDefined', {
              defaultMessage: 'Index options for {transformDetails} not defined',
              values: { transformDetails }
            })
        );
        return;
    }
  }

  const rawResults = await readFile(file);
  const jsonResult = parseAndClean(rawResults, cleanAndValidate)
  jsonPreview(jsonResult, previewCallback);

  return jsonResult;
}
```

Reply:
Slick. Works for me!